### ISSUE TITLE
Begin converting Language Server to go [phab:D1814]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2255]
_Created by **djrenren** at 2017-07-28 22:40:32_: 

This commit is still WIP. Some namings are wonky and tests are missing. Also diffing is currently omitted

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:46:51_:

Please remove all commented code.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:46:51_:

You can use sort.Slice

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:46:51_:

Fix copyright years to be 2017

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:46:51_:

Split import groups to be: kythe.io, protos, third_party

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:46:51_:

Use the log package.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:46:51_:

Fix trailing newlines throughout

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-07-28 22:46:51_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:48:08_:

Use a var declaration group:

```
var (
  root = flag.String...
  corpus = flag.String...
)
```

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-28 22:48:27_:

  - what-year

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:50:27_:

Split the arguments passed to this function to multiple lines (like multiline struct literals)

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 22:50:27_:

Remove extraneous newline.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 23:02:26_:

Don't rename imports unless forced (or /grumble/ proto... /grumble/)

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 23:02:26_:

Describe the binary package as: "Binary <name> <blah>..."

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 23:02:26_:

Document the package as: "Package languageserver <blah>..."

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 23:02:26_:

Rename to `CrossReferences` to match rest of platform.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-07-28 23:04:21_:

```
return kytheuri.URI{
  Corpus: p.Corpus,
  Path: loc,
}, nil
```

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-28 23:40:08_:

  - Add logging and fix style comments

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

Don't split up lines arbitrarily; Go doesn't assume Hollerith cards. As a rule of thumb, don't force code to any particular fixed width; break where it is semantically comfortable (and here, not at all).

    return "", fmt.Errorf("Path %q does not start with root %q", u.Path, u.Root)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

Don't ignore the possible error. While it's unlikely, the circumstances under which the process can't get its cwd would be serious enough to deserve a diagnostic.

```
var wd string

func init() {
	var err error
	wd, err = os.Getwd()
	if err != nil {
		log.Fatalf("Unable to find working directory: %v", err)
	}
}
```

or words to that effect

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

This is probably not the behaviour you want; close here should probably only close the downstream channel, not the upstream. Otherwise it is likely to result in double-closes.

I think this would be simpler as:

```
type stdio struct{}

func (stdio) Read(data []byte) (int, error)  { return os.Stdin.Read(data) }
func (stdio) Write(data []byte) (int, error) { return os.Stdout.Write(data) }
func (stdio) Close() error                   { return os.Stdout.Close() }
```

since that's the only way it gets used here anyway.

Then the above call can more simply be `jsonrpc2.NewBufferedStream(stdio{}, jsonrpc2.VSCodeObjectCodec{})`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

Why is this type here? Can you not use `lsp.Range` directly for some reason?

The methods below could be more simply free functions, e.g.,

    func rangeContains(range *lsp.Range, pos *lsp.Position) bool { return posLess(…) }

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

Why pointers? It looks like they're usually handled by value upstream, and these methods don't modify them.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

If you're going to split this anyway, try

```
if a.Line == b.Line {
    return a.Character < b.Character
}
return a.Line < b.Line
```

Otherwise, this can be a simple one-liner:

```
func posLess(a, b lsp.Position) bool {
	return a.Line < b.Line || (a.Line == b.Line && a.Character < b.Character)
}
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

cf. line 98

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

Don't ignore errors.

```
var p lsp.DidOpenTextDocumentParams
if err := json.Unmarshal(*req.Params, &p); err != nil {
  return nil, err
}
return nil, ls.TextDoumentDidOpen(p)
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-07-30 00:35:33_:

Although you may not need them here, I recommend giving these methods leading context.Context parameters. Most anything that touches the network (including via jsonrpc2) will wind up needing it to carry cancellations, deadlines, auth tokens, and so forth.

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-31 18:19:24_:

  - Address readability feedback

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-31 20:27:14_:

  - Add some unit tests

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-07-31 23:27:40_:

  - Add integration test for languageserver

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-01 18:40:11_:

Add flags for the port and hostname

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-01 18:40:11_:

Ping the server to allow for quick feedback if it's down.  A good ping might be a Decorations call with an empty request.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-01 18:40:11_:

Add a link to the spec and comment what exactly is implemented.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-01 18:40:11_:

Scratch that, just use kythe.io/kythe/go/services/xrefs.WebClient and kythe.io/kythe/go/services/xrefs.Service.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-01 18:40:11_:

Longer lines are okay in Go.  No need to wrap this call.

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-08-01 18:40:11_

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-01 18:56:13_:

  - Fix doc string for corpus option

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-01 21:15:11_:

  - Address feedback

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-01 22:17:50_:

  - Remove unnecessary linewrap

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-08-01 23:33:27_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-02 16:48:22_:

Please split the import groups throughout

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-02 16:48:22_:

`return fmt.Errorf("error creating local path from URI %s: %v", params.TextDocument.URI, err)`

Let the caller log the entire error.  Change this for all of the log/return err pairings.

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-02 20:26:06_:

  - Group exports and move error logging out of languageserver

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-08-02 20:31:44_:

Error messages should start with a lowercase character.

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-02 20:36:39_:

  - Lowercase errors are a thing I guess

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2017-08-02 20:38:15_

**Comment by kythe** [comment]
_**djrenren** closed via commit 62af21b6a69ad7b147dcea44ba97871e8933ef52 at 2017-08-02 20:47:46_
